compiler/noirc_evaluator/src/brillig/brillig_ir/codegen_stack.rs

```diff
@@ -1,122 +1,152 @@
-    /// The only requirement is that every destination needs to be written at most once.
+    /// It assumes that:
+    /// - every destination needs to be written at most once. Will panic if not.
+    /// - destinations are relative addresses, starting from 0 and with 1 increment. Will panic if not
+    /// - sources are any relative addresses, and can also be direct address to the global space. TODO: panic if not
+    /// - sources and destinations have same length. Will panic if not.
     pub(crate) fn codegen_mov_registers_to_registers(
         &mut self,
         sources: Vec<MemoryAddress>,
         destinations: Vec<MemoryAddress>,
     ) {
         assert_eq!(sources.len(), destinations.len());
-        // Remove all no-ops
-        let movements: Vec<_> = sources
-            .into_iter()
-            .zip(destinations)
-            .filter(|(source, destination)| source != destination)
-            .collect();
-
-        // Now we need to detect all cycles.
-        // First build a map of the movements. Note that a source could have multiple destinations
-        let mut movements_map: HashMap<MemoryAddress, HashSet<_>> =
-            movements.into_iter().fold(HashMap::default(), |mut map, (source, destination)| {
-                map.entry(source).or_default().insert(destination);
-                map
-            });
-
-        let destinations_set: HashSet<_> = movements_map.values().flatten().copied().collect();
-        assert_eq!(
-            destinations_set.len(),
-            movements_map.values().flatten().count(),
-            "Multiple moves to the same register found"
-        );
-
-        let mut loop_detector = LoopDetector::default();
-        loop_detector.collect_loops(&movements_map);
-        let loops = loop_detector.loops;
-
-        // In order to break the loops we need to store one register from each in a temporary and then use that temporary as source.
-        let mut temporaries = Vec::with_capacity(loops.len());
-
-        for loop_found in loops {
-            let temp_register = self.allocate_register();
-            let first_source = loop_found.iter().next().unwrap();
-            self.mov_instruction(*temp_register, *first_source);
-            let destinations_of_temp = movements_map.remove(first_source).unwrap();
-            movements_map.insert(*temp_register, destinations_of_temp);
-            temporaries.push(temp_register);
+        let n = sources.len();
+        let start = Stack::start();
+        let mut free_register = None;
+        let mut processed = 0;
+        // Compute the number of children for each node in the movement graph
+        let mut children = vec![0; n];
+        for i in 0..n {
+            // Check that destinations are relatives to 0,..,n-1
+            assert_eq!(destinations[i].unwrap_relative() - start, i);
+            let index = Self::to_index(&sources[i]);
+            if index < n && index != i {
+                children[index] += 1;
```

Comment:
I'm trying to understand this: say I want to copy 5 source registers to 5 destinations, the destinations are 1,2,3,4,5, and my first source is `Relative(100)`. The `index` here is going to be 99, and we won't be recording anything in `children`. Is it because `children` only contains the ones which overlap with the destinations? If so, can we mention this in the comment about `children`? I thought it was the number of destinations the source is copied to.

Reply:
yes it is (the answer to all your questions).
I'll add a comment to say that children is the number of destinations the source is copied to, for 'overlapping' sources.